Matter manipulator render

### Your GTNH Discord Username

mamiemru

### Your Pack Version

daily GTNH-2025-12-13+282-mmcprism-new-java

### Your Server

SP

### Java Version

Java 25

### Type of Server

Single Player

### Your Actions

The game crash when using matter manipulator 3.

steps to reproduce:
- Create a world in creative
- go to pdim no weather flat plain day only no tree ....
- i extended the main platforme with builder wand and built a multiblock with projector
- then i gave myself the charged MM mk3 from NEI
- When open the inventory the game crash twice.



### Crash Report

<img width="631" height="83" alt="Image" src="https://github.com/user-attachments/assets/bbe8e862-f857-45b1-8b60-6dcdda38fdc0" />



`---- Minecraft Crash Report ----
// Hi. I'm Minecraft, and I'm a crashaholic.

Time: 2025-12-14 10:53:58 CET
Description: Unexpected error

com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_ARRAY but was STRING at path $
    at com.google.gson.Gson.fromJson(Gson.java:1238)
    at com.google.gson.Gson.fromJson(Gson.java:1329)
    at com.google.gson.Gson.fromJson(Gson.java:1271)
    at com.recursive_pineapple.matter_manipulator.common.items.manipulator.MMState.migrateJson(MMState.java:159)
    at com.recursive_pineapple.matter_manipulator.common.items.manipulator.MMState.load(MMState.java:114)
    at com.recursive_pineapple.matter_manipulator.common.items.manipulator.ItemMatterManipulator.getState(ItemMatterManipulator.java:399)
    at com.recursive_pineapple.matter_manipulator.common.items.manipulator.MMKeyInputs.onKeyPressed(MMKeyInputs.java:45)
    at com.gtnewhorizon.gtnhlib.eventbus.StaticASMEventHandler_40_MMKeyInputs_onKeyPressed_InputEvent$KeyInputEvent.invoke(.dynamic)
    at com.gtnewhorizon.gtnhlib.eventbus.StaticASMEventHandler.invoke(StaticASMEventHandler.java:39)
    at cpw.mods.fml.common.eventhandler.EventBus.post(EventBus.java:140)
    at cpw.mods.fml.common.FMLCommonHandler.fireKeyInput(FMLCommonHandler.java:540)
    at net.minecraft.client.Minecraft.runTick(Minecraft.java:1873)
    at net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:973)
    at net.minecraft.client.Minecraft.run(Minecraft.java:7610)
    at net.minecraft.client.main.Main.main(SourceFile:148)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke(Method.java:565)
    at net.minecraft.launchwrapper.Launch.rfb$realLaunch(Launch.java:250)
    at net.minecraft.launchwrapper.Launch.launch(Launch.java:35)
    at net.minecraft.launchwrapper.Launch.main(Launch.java:60)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke(Method.java:565)
    at com.gtnewhorizons.retrofuturabootstrap.Main.main(Main.java:207)
    at com.gtnewhorizons.retrofuturabootstrap.MainStartOnFirstThread.lambda$main$1(MainStartOnFirstThread.java:41)
    at java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.IllegalStateException: Expected BEGIN_ARRAY but was STRING at path $
    at com.google.gson.internal.bind.JsonTreeReader.expect(JsonTreeReader.java:165)
    at com.google.gson.internal.bind.JsonTreeReader.beginArray(JsonTreeReader.java:73)
    at com.google.gson.internal.bind.TypeAdapters$2.read(TypeAdapters.java:90)
    at com.google.gson.internal.bind.TypeAdapters$2.read(TypeAdapters.java:87)
    at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:204)
    at com.google.gson.Gson.fromJson(Gson.java:1227)
    ... 24 more
`

### Final Checklist

- [x] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying the crash still exists will prompt us to investigate and reopen it once we confirm your report.
- [x] I can reproduce this crash consistently by following the exact steps I described above.
- [ ] I have asked other people and they confirm they also crash by following the exact steps I described above.

--- a/src/main/java/com/recursive_pineapple/matter_manipulator/common/items/manipulator/MMState.java
+++ b/src/main/java/com/recursive_pineapple/matter_manipulator/common/items/manipulator/MMState.java
@@ -84,9 +84,9 @@ public class MMState {
         .create();
 
     @SerializedName("jv")
-    private int jsonVersion = 0;
+    private int jsonVersion = LASTEST_JSON_VERSION;
     @SerializedName("dv")
-    private int dataVersion = 0;
+    private int dataVersion = LASTEST_DATA_VERSION;
 
     public MMConfig config = new MMConfig();
 
@@ -138,8 +138,11 @@ public class MMState {
         return copy;
     }
 
+    private static final int LASTEST_JSON_VERSION = 2;
+    private static final int LASTEST_DATA_VERSION = 0;
+
     private static void migrateJson(JsonObject obj) {
-        int version = obj.has("jv") ? obj.get("jv").getAsInt() : 2;
+        int version = obj.has("jv") ? obj.get("jv").getAsInt() : LASTEST_JSON_VERSION;
 
         if (version == 0) {
             if (obj.get("config") instanceof JsonObject config) {
@@ -157,11 +160,13 @@ public class MMState {
 
         if (version == 1) {
             try {
-                // Load with the default encoder
-                BitSet bitSet = new Gson().fromJson(obj.get("installedUpgrades"), BitSet.class);
+                if (obj.get("installedUpgrades") != null && obj.get("installedUpgrades").isJsonArray()) {
+                    // Load with the default encoder
+                    BitSet bitSet = new Gson().fromJson(obj.get("installedUpgrades"), BitSet.class);
 
-                // Save with the new encoder
-                obj.add("installedUpgrades", GSON.toJsonTree(bitSet));
+                    // Save with the new encoder
+                    obj.add("installedUpgrades", GSON.toJsonTree(bitSet));
+                }
             } catch (Throwable t) {
                 MMMod.LOG.error("Could not migrate installedUpgrades: your MM upgrades may have been deleted.", t);
             }

